Lib/test/libregrtest/refleak.py

```diff
@@ -145,6 +145,11 @@ def get_pooled_int(value):
             # Use an internal-only keyword argument that mypy doesn't know yet
             _only_immortal=True)  # type: ignore[call-arg]
         alloc_after = getallocatedblocks() - interned_immortal_after
+        if _get_tlbc_blocks := getattr(sys, "_get_tlbc_blocks", None):
```

Comment:
I don't think it will be practical to modify the tests so that they pass reliably in regrtest mode and it will create a confusing experience for people who add new tests. The issue is that new copies of the bytecode are created the first time a thread calls a function. In tests that use multiple threads it's possible that a different thread ends up calling some function each time the test is run. This results in a new copy of the bytecode being created for each test run, which regrtest treats as a leak.

More concretely, imagine we have a test using four threads, and regrtest is running 3 warmup iterations and 3 test iterations, and we encounter the following sequence:

1. Function F is called by thread 0 in the 3 warmup iterations. This will create one copy of the bytecode.
2. F is called by thread 1 in the first test iteration. This will create one copy of the bytecode. regrtest will treat this as a leaked object.
3. F is called by thread 2 in the second test iteration. This creates another copy of the bytecode that is treated as a leak.
4. F is called by thread 3 in the third test iteration.  This creates another copy of the bytecode that is treated as a leak.

To make this pass reliably we'd need to either ensure that all of threads create their copies of the bytecode in the warmup runs.

Since what we really care about is checking that bytecode copies are cleaned up, I can write an isolated test for that.

Reply:
We have similar issues with the JIT, which allocates blocks at hard-to-predict times.
If a test is using multiple threads, then they will have all run in the first iteration of the test. The refleak tests do not spawn additional threads (I think).

How many tests are failing without this?

Reply:
> If a test is using multiple threads, then they will have all run in the first iteration of the test. The refleak tests do not spawn additional threads (I think).

The issue is not related to thread creation, it's related to when threads call functions for the first time (which requires creating a new copy of the bytecode). If a function is only called once during the duration of a test (e.g. `notify` on a condition variable), then it's possible that a different thread will end up calling the function each time the test is run. This will appear as a leak in the refleak tests, since each run of the test in this scenario will result in a new copy of the bytecode being created.

> How many tests are failing without this?

It's a relatively small number of test suites, but they fail intermittently.

Reply:
Do we not clean up the copy of the bytecode when the thread exits?

Reply:
> Do we not clean up the copy of the bytecode when the thread exits?

No, they remain in the array attached to the code object for reuse by the next thread that is assigned the same tlbc index.

Reply:
I still think we need to count the blocks, otherwise we cannot spot regressions. 

Can we modify the failing tests somehow to be a bit more predictable in how many threads they use?
Or increase the number of iterations to stabilize those tests?

Reply:
> Can we modify the failing tests somehow to be a bit more predictable in how many threads they use?

The issue isn't the number threads that a test uses, per-se. It's whether or not a thread calls a function for the first time during the execution of a test.

> Or increase the number of iterations to stabilize those tests?

This might be possible, but there's no guarantee that it would work, and it would require increasing the number of iterations for *all* tests that are run during refleak tests. AFAIK there's no way to manually specify the number of iterations required for a specific test. (This could be added, of course.)

> I still think we need to count the blocks, otherwise we cannot spot regressions.

Agreed that we want to catch regressions. I added a [test](https://github.com/python/cpython/pull/123926/files#diff-01d18615b7d1791994f2cd69f42a07b9c9fdb07b7c0081477ec925e12e2b3701R195-R230) to verify that copies of the bytecode are cleaned up when code objects are destroyed.